File: src/zig/src/Launcher.zig
const std = @import("std");
const config = @import("config");
const Config = @import("Cfg.zig").Config;
const c = @cImport({
    @cInclude("shlwapi.h");
    @cInclude("shlobj.h");
});

const sound_config_file = switch (config.game_type) {
    .ff7 => "ff7sound.cfg",
    .ff8 => "ff8sound.cfg",
};
const video_config_file = switch (config.game_type) {
    .ff7 => "ff7video.cfg",
    .ff8 => "ff8video.cfg",
};

fn processGameMessages() u32 {
    std.debug.print("Starting game message queue thread...", .{});
}

fn getDocumentPath() !std.ArrayList(u8) {
    const gpa = std.heap.page_allocator;
    var doc_path = std.ArrayList(u8).init(gpa);

    const data_music_2 = std.fs.cwd().statFile("data/music_2") catch null;
    if (data_music_2 == null) { //TODO add ff7_estore_edition check
        var out_path: [256]u16 = undefined;
        const hr = c.SHGetKnownFolderPath(&c.FOLDERID_Documents, c.KF_FLAG_DEFAULT, null, @ptrCast(@alignCast(&out_path)));
        const path = try std.unicode.utf16LeToUtf8Alloc(gpa, &out_path);
        defer gpa.free(path);

        if (c.SUCCEEDED(hr)) {
            try doc_path.appendSlice(path);
            try doc_path.appendSlice("\\Square Enix\\FINAL FANTASY");
            switch (config.game_type) {
                .ff7 => try doc_path.appendSlice("VII Steam"),
                .ff8 => try doc_path.appendSlice("VIII Steam"),
            }
        }
    } else {
        _ = try std.fs.cwd().realpath("", doc_path.items);
    }

    return doc_path;
}

pub fn writeVideoConfig(cfg: Config) !void {
    var filepath = try getDocumentPath();
    defer filepath.deinit();

    try filepath.appendSlice(video_config_file);
    const file = try std.fs.cwd().openFile(filepath.items, .{ .mode = .read_write });
    defer file.close();
    try file.writer().writeInt(i32, @as(i32, cfg.window_width), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.window_height), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.refresh_rate), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.fullscreen), .little);
    try file.writer().writeInt(i32, 0, .little);
    try file.writer().writeInt(i32, @as(i32, cfg.keep_aspect_ratio), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.enable_linear_filtering), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.original_mode), .little);
    if (config.game_type == .ff8) {
        try file.writer().writeInt(i32, @as(i32, cfg.pause_game_on_background), .little);
    }
}

pub fn writeSoundConfig(cfg: Config) !void {
    var filepath = try getDocumentPath();
    defer filepath.deinit();

    try filepath.appendSlice(sound_config_file);
    const file = try std.fs.cwd().openFile(filepath.items, .{ .mode = .read_write });
    defer file.close();
    try file.writer().writeInt(i32, @as(i32, cfg.sfx_volume), .little);
    try file.writer().writeInt(i32, @as(i32, cfg.music_volume), .little);
}

File: ff78-launcher/src/launcher.rs
use std::io::Write;

use anyhow::Result;
use windows::Win32::{
    System::Com::CoTaskMemFree,
    UI::Shell::{FOLDERID_Documents, SHGetKnownFolderPath, KF_FLAG_DEFAULT},
};

use crate::{Context, GameType, StoreType};

const FF7_USER_SAVE_DIR: u32 = 10;
const FF7_DOC_DIR: u32 = 11;
const FF7_INSTALL_DIR: u32 = 12;
const FF7_LOCALE_DATA_DIR: u32 = 13;
const FF7_GAME_VERSION: u32 = 18;
const FF7_DISABLE_CLOUD: u32 = 22;
const FF7_END_USER_INFO: u32 = 24;

const FF8_USER_SAVE_DIR: u32 = 9;
const FF8_DOC_DIR: u32 = 10;
const FF8_INSTALL_DIR: u32 = 11;
const FF8_LOCALE_DATA_DIR: u32 = 12;
const FF8_GAME_VERSION: u32 = 17;
const FF8_DISABLE_CLOUD: u32 = 21;
const FF8_BG_PAUSE_ENABLED: u32 = 23;
const FF8_END_USER_INFO: u32 = 24;

const ESTORE_USER_SAVE_DIR: u32 = 9;
const ESTORE_DOC_DIR: u32 = 10;
const ESTORE_INSTALL_DIR: u32 = 11;
const ESTORE_LOCALE_DATA_DIR: u32 = 12;
const ESTORE_GAME_VERSION: u32 = 17;
const ESTORE_END_USER_INFO: u32 = 20;

static mut LAUNCHER_MEMORY_PART: Vec<u8> = Vec::new();

pub fn get_game_install_path(ctx: &Context) -> Result<String> {
    let mut game_install_path = String::new();
    if !matches!(ctx.game_to_launch, GameType::FF7(StoreType::EStore))
        && std::fs::exists("data/music_2").is_err()
    {
        let doc_path = unsafe {
            let doc_path_pw = SHGetKnownFolderPath(&FOLDERID_Documents, KF_FLAG_DEFAULT, None)?;
            let doc_path = doc_path_pw.to_string()?;
            CoTaskMemFree(Some(doc_path_pw.as_ptr() as *const std::ffi::c_void));
            doc_path
        };
        game_install_path += &doc_path;
        game_install_path += "\\Square Enix\\FINAL FANTASY ";
        game_install_path += match ctx.game_to_launch {
            GameType::FF7(_) => "VII Steam",
            GameType::FF8 => "VIII Steam",
        }
    } else {
        let cwd = std::env::current_dir()?
            .to_str()
            .ok_or(anyhow::anyhow!("cwd cannot be converted to string"))?
            .to_string();
        game_install_path += &cwd;
    }
    Ok(game_install_path)
}

pub fn send_locale_data_dir(ctx: &Context) {
    let mem_payload = String::from("lang-") + &ctx.game_lang;
    let mut launcher_game_part = Vec::<u8>::new();
    launcher_game_part.push(match ctx.game_to_launch {
        GameType::FF7(StoreType::Standard) => FF7_LOCALE_DATA_DIR as u8,
        GameType::FF7(StoreType::EStore) => ESTORE_LOCALE_DATA_DIR as u8,
        GameType::FF8 => FF8_LOCALE_DATA_DIR as u8,
    });
    launcher_game_part.append(&mut (mem_payload.len() as u32).to_le_bytes().to_vec());
    launcher_game_part.append(
        &mut mem_payload
            .encode_utf16()
            .flat_map(|b| b.to_le_bytes())
            .collect(),
    );
    launcher_game_part.push(0);
    unsafe {
        // NOTE: Unsafe since single threaded
        LAUNCHER_MEMORY_PART = launcher_game_part;
    };
    log::info!("send_locale_data_dir -> {mem_payload}");

    // TODO:
}

pub fn write_ffvideo(ctx: &Context) -> Result<()> {
    let filename = match ctx.game_to_launch {
        GameType::FF7(_) => "ff7video.cfg",
        GameType::FF8 => "ff8video.cfg",
    };
    let filepath = get_game_install_path(ctx)? + "\\" + filename;
    let mut file = std::fs::File::create(filepath)?;
    file.write_all(&ctx.config.window_width.to_le_bytes())?;
    file.write_all(&ctx.config.window_height.to_le_bytes())?;
    file.write_all(&ctx.config.refresh_rate.to_le_bytes())?;
    file.write_all(&u32::from(ctx.config.fullscreen).to_le_bytes())?;
    file.write_all(&0u32.to_le_bytes())?;
    file.write_all(&u32::from(ctx.config.keep_aspect_ratio).to_le_bytes())?;
    file.write_all(&u32::from(ctx.config.enable_linear_filtering).to_le_bytes())?;
    file.write_all(&u32::from(ctx.config.original_mode).to_le_bytes())?;
    if let GameType::FF8 = ctx.game_to_launch {
        file.write_all(&u32::from(ctx.config.pause_game_on_background).to_le_bytes())?;
    }
    Ok(())
}

pub fn write_ffsound(ctx: &Context) -> Result<()> {
    let filename = match ctx.game_to_launch {
        GameType::FF7(_) => "ff7sound.cfg",
        GameType::FF8 => "ff8sound.cfg",
    };
    let filepath = get_game_install_path(ctx)? + "\\" + filename;
    let mut file = std::fs::File::create(filepath)?;
    file.write_all(&ctx.config.sfx_volume.to_le_bytes())?;
    file.write_all(&ctx.config.music_volume.to_le_bytes())?;
    Ok(())
}
